Add EIP: Theorem-Carrying-Transaction - TCT (#7837)
ethereum/EIPs: EIPS/eip-7534.md

---
eip: 7534

> title: TCT

Reviewer: ```suggestion
title: Theorem-Carrying-Transaction - TCT
```

> description: TCT combines the benefits of runtime checking and symbolic proof and ensures the theorems are provable and checkable in an efficient manner.

Reviewer: should describe what TCT is, not the benefits which can go into motivation

Author: I have updated the description.

author: Shuo Chen (@cs0317), Nikolaj Bjørner (@NikolajBjorner), Tzu-Han Hsu (@tzuhancs), Ashley Chen (@ash-jyc), Nanqing Luo (@Billy1900)
discussions-to: https://ethereum-magicians.org/t/eip-7534-transactioncarrying-theorem-tct-proposal/17125
status: Draft
type: Standards Track
category: Core
created: 2023-10-12
---

## Abstract

> Smart contracts are crucial elements of decentralized technologies, but they face significant obstacles to trustworthiness due to security bugs and trapdoors. To address the core issue, we propose a technology that enables programmers to focus on design-level properties rather than specific low-level attack patterns. Our proposed technology, called Theorem-Carrying-Transaction (TCT), combines the benefits of runtime checking and symbolic proof. Under the TCT protocol, every transaction must carry a theorem that proves its adherence to the safety properties in the invoked contracts, and the blockchain checks the proof before executing the transaction. The unique design of TCT ensures that the theorems are provable and checkable in an efficient manner. We believe that TCT holds a great promise for enabling provably secure smart contracts in the future.

Reviewer: ```suggestion
Smart contracts are crucial elements of decentralized technologies, but they face significant obstacles to trustworthiness due to security bugs and trapdoors. To address the core issue, we propose a technology that enables programmers to focus on design-level properties rather than specific low-level attack patterns. Our proposed technology, called Theorem-Carrying-Transaction (TCT), combines the benefits of runtime checking and symbolic proof. Under the TCT protocol, every transaction must carry a theorem that proves its adherence to the safety properties in the invoked contracts, and the EL clients checks the proof before executing the transaction. The unique design of TCT ensures that the theorems are provable and checkable in an efficient manner. We believe that TCT holds a great promise for enabling provably secure smart contracts in the future.
```

## Specification

First, we give several terminology:

- $\tau$: it denotes the theorem. Every transaction must carry a theorem that proves its adherence to the specified safety properties in the invoked contracts.
  
- $\mathrm{f}$: it denotes the entry function of transaction.
  
- $\varphi(\mathrm{V}, \mathrm{s})$: $\varphi$ denotes hypothesis which defined over input parameter $\mathrm{V}$ and blockchain state $\mathrm{s}$.
  
- $\mathrm{h}$: code path hash which is used for execution path match.
  

> - Invariant: contract invariants like the ones are ensured before and after every transaction. They contain quantifiers, sum of map, etc., so are difficult to check concretely.

Reviewer: ```suggestion
- Invariant: contract invariants like the ones are ensured before and after every transaction. They contain quantifiers, sum of map, etc., which are difficult to check concretely.
```
?

  
- behavioral subtyping: Unlike the conventional notion of OOP subtyping, which underpins today’s smart contracts and many other real-world systems, behavioral subtyping ensures that a derived class cannot violate properties defined in a base class.

For each transaction which is executed, it SHOULD satisfies the following theorem:

$$
\tau::=(\mathrm{f}, \varphi(\mathrm{V}, \mathrm{s}), \mathrm{h})
$$

The above theorem means​ for any transaction started by invoking $\mathrm{f}$ when $\varphi$ is satisfied, if it is completed (i.e., not reverted by EVM) and the hash of the code path equals $\mathrm{h}$, then all the assertions along the code path are guaranteed to hold. ​

Technically, code path hash $\mathrm{h}$ will be added into `Transaction` struct so each transaction will carry its own hypothesis hash $\mathrm{h}$.

```go

> // Transaction is an Ethereum transaction.

Reviewer: doesn't seem to follow ethereum transaction specification

Author: do you mean we modify the `Transaction` struct which violates the ethereum transaction specification?

    result, err := ApplyMessage(evm, msg, gp)
    ...
}
```

## Rationale

Conventional approaches predominantly resort to specialized tools and the expertise of domain professionals to detect and address known vulnerabilities or bugs, targeting the prevention of unintended contract behaviors. While these methods have their merits, empirical evidence suggests a palpable shortfall in ensuring comprehensive safety features. In contrast, our application of TCT brings forth a robust security guarantee via runtime verification, markedly elevating the safety standards compared to the traditional strategies.  

While numerous projects delve into runtime verification, our approach is distinctly tailored to focus solely on transactions which has several merits:

- **Minimal runtime overhead**: By so doing, we anticipate a tangible reduction in the overhead typically associated with blanket verification mechanisms.

- **Concolic testing**: Further, by synthesizing the merits of runtime verification with symbolic proof, our methodology stands out in ensuring both the provability and verifiability of theorems, setting a benchmark in efficacy and reliability. 

## Backwards Compatibility

> The primary concern lies in the potential for incompatibility, particularly with regards to the occurrence of a fork. The Theorem-Carrying-Transaction (TCT) system may reject certain transactions that violate the theorem, while the original Ethereum network may accept these transactions. This discrepancy in transaction acceptance could lead to a network fork, resulting in divergent block states. To address this issue, we propose conducting a comprehensive evaluation of TCT's effectiveness and stability by initially deploying it on a testnet such as `Goeril`.

Reviewer: if its a new transaction type enabled at some fork i don't understand how it makes it backward incompatible

Author: You are right, it will be incompatible. That's why we want to use a testnet to test TCT first.
